libarchive/archive_read_disk_entry_from_file.c

```diff
@@ -419,12 +419,28 @@ setup_acls(struct archive_read_disk *a,
 	if (accpath == NULL)
 		accpath = archive_entry_pathname(entry);
 
+	if (*fd < 0 && a->tree != NULL) {
+		if (a->follow_symlinks ||
+		    archive_entry_filetype(entry) != AE_IFLNK)
+			*fd = a->open_on_current_dir(a->tree,
+			    accpath, O_RDONLY | O_NONBLOCK);
+		if (*fd < 0) {
+			if (a->tree_enter_working_dir(a->tree) != 0) {
+				archive_set_error(&a->archive, errno,
+				    "Couldn't access %s", accpath);
+				return (ARCHIVE_FAILED);
+			}
+		}
+	}
+
 	archive_entry_acl_clear(entry);
 
+	acl = NULL;
+
 #ifdef ACL_TYPE_NFS4
 	/* Try NFS4 ACL first. */
 	if (*fd >= 0)
-		acl = acl_get_fd(*fd);
+		acl = acl_get_fd_np(*fd, ACL_TYPE_NFS4);
```

Comment:
It has been a long time since I looked at Linux' NFS4 support, but I don't remember them trying to merge POSIX.1e and NFS4 ACLs in the same way that FreeBSD did.

However, I do remember looking at one platform that went even further and allowed mixing NFS4 and POSIX.1e ACLs on the same filesystem at the same time.  They must have had some API for distinguishing them.